### DISCUSSION TITLE
INSERT INTO parquet file ?

### DISCUSSION THREAD

**Discussion by bladko-tom** [duckdb/duckdb#7547] [Ideas]
We know to use COPY FROM / COPY TO to operate to read / write parquet files. INSERT INTO is at the moment only for tables within DB. 

I wonder if logic to actually insert into specific parquet file would be possible or useful. Or what would be potential alternative ?

Use case: Incremental ELT with daily increment that appended to parquet without use of hive partition  as some new records will replace existing one -> typical case for insert or replace on table.

**Comment by bartvdbraak** [comment]
Hey @bladko-tom!

While DuckDB currently supports reading and writing Parquet files using `COPY FROM` and `COPY TO`, the `INSERT INTO` statement is limited to tables within the database. However, I can suggest an alternative approach to achieve your desired use case of incremental ELT with daily increments appended to a Parquet file.

Instead of using `INSERT INTO`, you can leverage DuckDB's capability to directly read and write Parquet files. Here's a step-by-step outline you can follow:

1.    Start by using the `COPY FROM` statement to read the existing Parquet file into DuckDB. This will create a temporary table with the Parquet file's contents.

2.    Apply any necessary transformations or filters on the data within the temporary table.

3.    Next, use `COPY FROM` again to read the new daily increment data from your source (such as another Parquet file or CSV) into another temporary table within DuckDB.

4.    Apply any required transformations or filters on the new data within the temporary table.

5.    Now, you can use DuckDB's SQL syntax to combine the existing data from the first temporary table with the new data from the second temporary table. For example, you can use the `UNION ALL` or `UNION DISTINCT` operator to merge the two datasets.

6.    Finally, use the `COPY TO` statement to write the combined result back to the existing Parquet file. This action will overwrite the file with the updated data.

By following this approach, you'll be able to achieve incremental updates to your Parquet file without relying on Hive partitions. Remember to adjust the steps according to your specific requirements and data sources.

I hope this helps! Let me know if you have any further questions or need additional assistance.

**Reply by bladko-tom** [reply]
Thanks @bartvdbraak . Do you think this should work on 5GB parquet ? I mean imagine daily increment is like 100MB, current size of parquet is few GB and total number of tables is around 100.

Point is INSERT OR REPLACE into must be used (some transactions updates in increment data that exist already in master data) but due to sheer size of data its ultra slow, that is why i am asking. Saving to back to hive will not work as i have to update or insert new data that exist in daily increment. Or maybe Duck DB is smart enough to handle it ?

On the other hand if this is put into DuckDB db (storage ver 51) then after using several insert or replace DB grow to size much bigger then org parquet files. 

I want to use DuckDB for OLAP, it may have other technology for ETL..

**Reply by djouallah** [reply]
delta lake python API, is starting to get good enough, it maybe an option, here is a simple example that append data 

https://github.com/djouallah/aemo_tracker/blob/main/writedelta.py

**Comment by l1t1** [comment]
parquet doesn't support update in nature, so you may use many parquet files partition by date.

**Comment by Mytherin** [comment]
You can't really append to a Parquet file like you can to a CSV file or a DuckDB file as the format is not designed for that. The format is designed to be written to once, after which the metadata is written at the end of the file. What is generally done in these scenarios is having a directory filled with Parquet files, where "appending" becomes writing a new Parquet file to that directory.

**Reply by bladko-tom** [reply]
But INSERT OR REPLACE into DuckDB (even the same data over and over again with primary key) making DB growing all the time so i am looking for alternative.

**Reply by rsund** [reply]
Is there some specific reason that you could not add new daily parquet files to the directory as akready suggested two times above? Using `FROM read_parquet('filename*.parquet')` should work just fine regardles of how many files you have.

**Reply by bladko-tom** [reply]
Yes there is a reason, allow me to explain how my data refresh works now.

1.  Daily / bi-daily extract from transactional system to CSV is being converted 1:1 to parquet with time stamp. This is where i do also perform some cleanup / data quality checks as i have no control over it on data source systems.
2. This data contains new records but also updates to existing ones, so i have to remove duplicates and just keep latest's values. In this case history of same e.g. ticket is not relevant for analytics.  
3. For this purpose i build a **stage** section that takes incremental parquet and INSERT OR REPLACE into DuckDB with primary key to remove not needed records. I had to use DB as its impossible to do it via parquet only. 
4. Then for analytics i am either using this DB 
5. or  exporting back to Parquet on **prod** section this time using hive partitioning, usually based on monthly timestamp. Parquet files then are used for second DB that has views (not tables) over parquet files. This way its faster then doing analytics over same data in DuckDB and for my users it really does not matter what is backend as both DBs has exactly the same schema and table names. 

This works fine with the only problem DuckDB is growing even when inserting same data (at least in my observation aprox double the original size of table, leaving half blocks free space). I am looking for reliable solution in which i will know that storage would be x2 due to nature of DB OR it will take 2 times longer due to having additional stage part for clean up etc.. I would not prefer to use other technology like PostgreSQL if its would be not mandatory to do so.

I also found that reading / writing parquet via DuckDB is much more performant for analytics then using it via pyarrow. In other words parquet created by DuckDB works much better when query via DuckDB then same data saved via pyarrow.

Maybe that is the case for **materialized views** ? Data refresh is done via night so it may take 50% more time or x2 storage i dont mind, the point is to have it ready for users during the day and they should have max performance / experience possible.

**Comment by bladko-tom** [comment]
Allow me to share additional observation regarding "appending to DuckDB file as alternative to append to parquet" (which is impossible ok)

1) 3 parquet files which comes from 3 daily increments, each around 50 MB
2) DuckDB table with primary key on single column (so also index)
3) INSERT OR REPLACE INTO for the first time works great and this gives aprox 700MB DB size (which is not my concern compared to Parquet) and time is aprox 10 sec (which is also not a problem here)
4) Now rerun the same INSERT OR REPLACE INTO will not effectively insert any data due to constrains but makes DB grow
5) Now the point is its growing to specific level (in this case its aprox 1.15 GB) and then it stops growing regardless of how many inserts were run

Every time after insert i do commit, run FORCE CHECKPOINT and VACUUM / VACUUM ANALYZE but nothing seems to explain this behaviors in point number 5)

I tried to play around with size of wal_autocheckpoint and also try preserve_insertion_order but looks like it has no impact on subject. 

Checked on 0.7.1 and 0.8

**Reply by rsund** [reply]
I think there has been discussion somewhere earlier that `CHECKPOINT;` need to be run in order to get rid of WAL so that freed space could be reused, but if the size of data grows, it will not fit to earlier space and new space will be used in the database file. But these things may change from version to other as the format may (and will) change until version 1.0.

I also think that DuckDB does not have `VACUUM;` that is something else than no op yet (or I have missed it) and the possible way to do similar clean-up is the use of `EXPORT` (maybe to parquet if you need compression) and then `IMPORT` to a new database. Maybe keeping a backup (parquet+sql from export) is a good idea to avoid potential problems in case of DuckDB version updates. If EXPORT-IMPORT is not working in a way that preserves the database contents unchanged, that will be probably worth of an issue.

And in principle you could overwrite existing parquets to get them "updated", but that might be a bit risky unless you are sure that the earlier versions will be needed.

**Comment by l1t1** [comment]
OLAP db usually  didn't support frequently DML, duckdb is easy to use enough.

**Reply by bladko-tom** [reply]
Thanks, i know. I made this test to simulate daily or bi daily inserts. I will never insert same data twice to same table, or even twice per day. But my test was to simulate long term operation to check if this will work fine.

**Comment by l1t1** [comment]
how about create a directory for each primary key
then you only need to replace the only file in each directory.

**Comment by l1t1** [comment]
of course too many files have their own problems too. see https://github.com/duckdb/duckdb/discussions/7463

**Comment by l1t1** [comment]
ref https://github.com/ClickHouse/ClickHouse/discussions/49416
